### PR TITLE
feat(ui5-tooling-transpile): generate d.ts and filter ts files

### PR DESCRIPTION
**Pull request by petermuessig**

### CODE REVIEW

**Comment by petermuessig** [comment]
@texttechne @akudev - here my suggestion to include the `d.ts` files and to exclude the `ts` files from the build result. The `ts` files will now always be excluded from the build result but the generation of the `d.ts` files can also be disabled with the setting `skipGenerateDTS`. If ok, I will merge and release tomorrow...

File: packages/ui5-tooling-transpile/lib/task.js [hunk 1]
```diff
@@ -31,54 +32,150 @@ module.exports = async function ({ workspace, dependencies, /* taskUtil,*/ optio
 	}
 
 	// TODO: should we accept the full glob pattern as param or just the file pattern?
-	const allResources = await workspace.byGlob(`/**/*${filePatternConfig}`);
-	if (config.transpileDependencies) {
-		// TODO: does transpileDependencies make sense for JavaScript files?
-		const depsResources = await dependencies.byGlob(`/**/*${filePatternConfig}`);
-		allResources.push(...depsResources);
-	}
+	let allResources = await workspace.byGlob(`/**/*${filePatternConfig}`);
```
Comment (texttechne) [inline packages/ui5-tooling-transpile/lib/task.js:35]: can be `const`

File: packages/ui5-tooling-transpile/lib/task.js [hunk 2]
```diff
@@ -31,54 +32,150 @@ module.exports = async function ({ workspace, dependencies, /* taskUtil,*/ optio
...
+				//traceResolution: true,
+			};
+
+			// Create a Program with an in-memory emit
+			const host = ts.createCompilerHost(options);
+			(host.getCurrentDirectory = function () {
+				return "";
+			}),
+				(host.fileExists = function (file) {
+					if (sourcesMap[file]) {
```
Comment (texttechne) [inline packages/ui5-tooling-transpile/lib/task.js:116]: can be shortened: 
```
return sourcesMap[file] ? true : fs.existsSync(file)
```

File: packages/ui5-tooling-transpile/lib/task.js [hunk 3]
```diff
@@ -31,54 +32,150 @@ module.exports = async function ({ workspace, dependencies, /* taskUtil,*/ optio
...
+						return true;
+					} else {
+						return fs.existsSync(file);
+					}
+				});
+			host.readFile = function (file) {
+				if (sourcesMap[file]) {
```
Comment (texttechne) [inline packages/ui5-tooling-transpile/lib/task.js:123]: can be shortened to one-liner

File: packages/ui5-tooling-transpile/lib/task.js [hunk 4]
```diff
@@ -31,54 +32,150 @@ module.exports = async function ({ workspace, dependencies, /* taskUtil,*/ optio
...
+					path: `${fileName}`,
+					string: content
+				});
+				workspace.write(dtsFile);
+			};
+			host.resolveModuleNames = function (moduleNames, containingFile) {
+				const resolvedModules = [];
+				for (const moduleName of moduleNames) {
+					// try to use standard resolution
+					let result = ts.resolveModuleName(moduleName, containingFile, options, {
```
Comment (texttechne) [inline packages/ui5-tooling-transpile/lib/task.js:141]: can be `const`

**Review by texttechne — APPROVED** [review]

**Comment by texttechne** [comment]
I'm a bit surprised that the source map integration is hard-coded instead of being configurable.
On the one hand, not all people want to publish their source code in this way; usually some will insist on obfuscation.
On the other hand, there are multiple options to generate different kinds of source maps with different impact on build time and debugging quality.

**Comment by petermuessig** [comment]
The source maps integration needs to be hard-coded in this case since we need to create additional files in the workspace - so there needs to be some glue code somewhere. But the decision whether source maps will be created is based on the configuration in the `.babelrc` file. If the source maps are activated there, the transpile function returns an object with the source map attached and this needs to be written into the workspace.

File: packages/ui5-tooling-transpile/lib/task.js (hunk 1 above)
Comment (petermuessig) [reply packages/ui5-tooling-transpile/lib/task.js:35]: 👍

File: packages/ui5-tooling-transpile/lib/task.js (hunk 2 above)
Comment (petermuessig) [reply packages/ui5-tooling-transpile/lib/task.js:116]: Damn, the auto completion of VSCode completely changed the code and put it into brackets... FYI: the reason why I have chosen those lengthy code was to be able to put break points inside and add additional logging - therefore I prefer to stick to the lengthy code. And the reason why I used classic functions is to be able to use the `arguments` when logging... 😉

File: packages/ui5-tooling-transpile/lib/task.js (hunk 4 above)
Comment (petermuessig) [reply packages/ui5-tooling-transpile/lib/task.js:141]: copy paste from here: https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API - not reworked all places properly... 😁

**Comment by texttechne** [comment]
Ahhh, configurable via `.babelrc`. Overlooked, but understood now 😄.

The improvements I've suggested are only minor in nature. I've tested this against ui5-mobx and it works as it should. Hence I already approved this PR (not that it makes any difference 😉).

**Comment by petermuessig** [comment]
Ok, I'll rename

**Pull request merged**